### PR TITLE
[#61] specify datatypes before writing to zarr

### LINKED ISSUE #61: zarr datatype error
I'm getting this datatype error when trying to write to zarr

```
ValueError: unable to infer dtype on variable None; xarray cannot serialize arbitrary Python objects
```

I have these package versions:
xarray 2022.3.0
zarr 2.11.1

### PR DESCRIPTION
**Pull request by jsadler2**
I had to specify datatypes for for site_id and date before writing to Zarr. The other columns were ok.

closes #61

### CODE REVIEW

**Review by lekoenig — COMMENTED** [review]
These changes look good to me. I have a couple minor questions, and also note that I'm not currently able to build these zarr targets unless I change the format of "string" used by pandas. Do you know what version of pandas you're using?

File: 2a_model/src/model_ready_data_utils.R [hunk 1]
```diff
@@ -27,14 +27,18 @@ write_df_to_zarr <- function(df, index_cols, out_zarr) {
...
+  py_df[["site_id"]] = py_df$site_id$astype("string")
   
   # set the index so that when we convert to an xarray dataset it is indexed properly
   py_df  <- py_df$set_index(index_cols)
+
   
   # convert to an xarray dataset
   ds <- py_df$to_xarray()
-  
   ds$to_zarr(out_zarr, mode = 'w')
```
[the remark this answers is not in the document]
Comment (jsadler2) [reply 2a_model/src/model_ready_data_utils.R:40]: Yeah. Pandas is for tabular data. Xarray is for multi-dimensional - it was originally designed to be an "in-memory" representation of a netcdf file. Then zarr is basically like a chunked netcdf file. So we first convert to xarray to get to multi-dimensional (date and site) and then from there we can write to zarr. We can't write zarr from pandas.

File: 2a_model/src/model_ready_data_utils.R [hunk 2]
```diff
@@ -53,5 +57,6 @@ subset_and_write_zarr <- function(df, out_zarr, sites_subset = NULL){
     if (!is.null(sites_subset)){
       df <- df %>% filter(site_id %in% sites_subset)
     }
-  write_df_to_zarr(df, c("site_id", "date"), out_zarr)
+  out_zarr <- write_df_to_zarr(df, c("site_id", "date"), out_zarr)
+  return(out_zarr)
```
[the remark this answers is not in the document]
Comment (jsadler2) [reply 2a_model/src/model_ready_data_utils.R:61]: Yeah. I think this was independent. I just realized that this needed to be passed as I was looking at the code.

File: 2a_model/src/model_ready_data_utils.R [hunk 3]
```diff
@@ -27,14 +27,18 @@ write_df_to_zarr <- function(df, index_cols, out_zarr) {
   
   # convert to a python (pandas) DataFrame so we have access to the object methods (set_index and to_xarray)
   py_df <- reticulate::r_to_py(df)
+  pd <- reticulate::import("pandas")
+  py_df[["date"]] = pd$to_datetime(py_df$date)
+  py_df[["site_id"]] = py_df$site_id$astype("string")
```
[the remark this answers is not in the document]
Comment (jsadler2) [reply 2a_model/src/model_ready_data_utils.R:32]: Ah weird. Yeah "str" worked for me. I think "string" is a new thing.

Comment (jsadler2) [inline 2a_model/src/model_ready_data_utils.R:32]:
```suggestion
  py_df[["site_id"]] = py_df$site_id$astype("str")

```

File: 2a_model/src/model_ready_data_utils.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (lekoenig) [reply 2a_model/src/model_ready_data_utils.R:41]: OK, thanks, that makes sense.

File: 2a_model/src/model_ready_data_utils.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (lekoenig) [reply 2a_model/src/model_ready_data_utils.R]: OK, that's great if it works and doesn't change any of the functionality. Eventually, we may want to add some information to the README that describes the python dependencies and perhaps even how to install them using `reticulate`.

**Review by lekoenig — APPROVED** [review]
Thanks for checking on that pandas "string" question. I haven't checked out the zarr files but the pipeline now builds for me, so feel free to merge whenever you're ready.

**Pull request merged**